tests/gtest/cppapi_test.cpp

```diff
@@ -0,0 +1,104 @@
+#include <map>
+
+#include "ctypes.h"
+#include "cppapi.hpp"
+
+#include "gtest/gtest.h"
+#include "gmock/gmock.h"
+
+namespace
+{
+    const std::string SAMPLES_10 = "file://10_samples_default.vds";
+    const std::string CREDENTIALS = "";
+
+    std::map<float, std::vector<float>> samples_10_data(float fill)
+    {
+        std::map<float, std::vector<float>> points;
+        points[110] = {-4.5, -3.5, -2.5, -1.5, -0.5, 0.5, 1.5, 2.5, 3.5, 4.5};
+        points[111] = {4.5, 3.5, 2.5, 1.5, 0.5, -0.5, -1.5, -2.5, -3.5, -4.5};
+        points[310] = {25.5, -14.5, -12.5, -10.5, -8.5, -6.5, -4.5, -2.5, -0.5, 25.5};
+        points[311] = {25.5, 0.5, 2.5, 4.5, 6.5, 8.5, 10.5, 12.5, 14.5, 25.5};
+        points[510] = {25.5, 4.5, 8.5, 12.5, 16.5, 20.5, 24.5, 20.5, 16.5, 8.5};
+        points[511] = {25.5, -4.5, -8.5, -12.5, -16.5, -20.5, -24.5, -20.5, -16.5, -8.5};
+        points[fill] = {fill};
+        return points;
+    }
+
+    class HorizonTest : public ::testing::Test
+    {
+    protected:
+        void SetUp() override
+        {
+            handle = make_datahandle(SAMPLES_10.c_str(), CREDENTIALS.c_str());
+        }
+
+        void TearDown() override
+        {
+            delete handle;
+        }
+
+        DataHandle *handle;
+
+        static constexpr float fill = -999.25;
+        std::map<float, std::vector<float>> points;
+    };
+
+    TEST_F(HorizonTest, DataForUnalignedSurface)
+    {
+        const float above = 2;
+        const float below = 2;
+
+        static constexpr int nrows = 4;
+        static constexpr int ncols = 6;
+        static constexpr std::size_t size = nrows * ncols;
+
+        const float xori = 2;
+        const float yori = 3;
+        const float xinc = 4.472;
+        const float yinc = 2.236;
+        const float rot = 333.43;
+
+        const float surface_data[nrows * ncols] = {
+            24, 20, 24, 24, 24, 20,
+            20, 20, 20, 24, 20, 24,
+            20, 24, 20, 20, 24, 20,
+            24, 24, 24, 24, 20, 24};
+
+        const float expected[nrows * ncols] = {
+            111, 311, 311, fill, fill, fill,
+            310, 310, 310, 311, 511, 511,
+            fill, fill, fill, fill, 510, 510,
+            fill, fill, fill, fill, fill, fill};
+
+        RegularSurface unaligned_surface = RegularSurface(
+            surface_data, nrows, ncols, xori, yori, xinc, yinc, rot, fill);
+
+        std::size_t horizon_size;
+        cppapi::horizon_size(*handle, unaligned_surface, above, below, &horizon_size);
+
+        float res[horizon_size];
```

Comment:
My decision process: 

Can I use std::array? Yes, because I know the size of it at compile time and I never need to change it
If no: use std::vector.

`cppapi::horizon` takes in a `void*` for the output buffer. This is makes it agnostic into how the buffer looks like, or what it's owned by (c array, std::array, std::vector, etc)

Both `std::vector` and `std::array` implement a member function `data()` which gives a pointer to the first element.
```
std::array< float, horizon_size >  res;
cppapi::horizon(*handle, unaligned_surface, above, below, NEAREST, 0, size, res.data());
```